journal/journal.go

```diff
@@ -0,0 +1,120 @@
+package journal
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/filecoin-project/go-filecoin/clock"
+)
+
+// Writer defines an interface for recording events and their metadata
+type Writer interface {
+	// Write records an operation and its metadata to a Journal accepting variadic key-value
+	// pairs.
+	Write(event string, kvs ...interface{})
+}
+
+// Journal defines an interface for creating Journals with a topic.
+type Journal interface {
+	// Topic returns a Writer that records events for a topic.
+	Topic(topic string) Writer
+}
+
+// NewInMemoryJournal returns a journal backed by an in-memory map.
+func NewInMemoryJournal(clk clock.Clock) Journal {
+	return &MemoryJournal{
+		clock:  clk,
+		topics: make(map[string]struct{}),
+	}
+}
+
+// MemoryJournal represents a journal held in memory.
+type MemoryJournal struct {
+	clock   clock.Clock
+	topics  map[string]struct{}
+	writers []*MemoryWriter
+}
+
+// Topic returns a Writer with the provided `topic`.
+func (mj *MemoryJournal) Topic(topic string) Writer {
+	if _, ok := mj.topics[topic]; ok {
+		// this means we have already made a journal here
+		// fail?
+		panic("duplicate topics developer error")
+	}
+	mr := &MemoryWriter{
+		clock: mj.clock,
+		topic: topic,
+	}
+	mj.writers = append(mj.writers, mr)
+	return mr
+}
+
+// String returns all Journal topics and their entries as a string.
+func (mj *MemoryJournal) String() string {
+	var sb strings.Builder
+	for _, writer := range mj.writers {
+		sb.WriteString(fmt.Sprintf("topic: %s", writer.topic))
+		sb.WriteRune('\n')
+		for _, ent := range writer.entries {
+			sb.WriteString(fmt.Sprintf("\t%s - %s: %v", ent.time.String(), ent.event, ent.kvs))
+			sb.WriteRune('\n')
+		}
+	}
+	return sb.String()
+}
+
+type entry struct {
+	time  time.Time
+	event string
+	kvs   map[string]interface{}
+}
+
+// MemoryWriter writes journal entires in memory.
+type MemoryWriter struct {
+	clock   clock.Clock
+	topic   string
+	entries []entry
+}
+
+// Write records an operation and its metadata to a Journal accepting variadic key-value
+// pairs.
+func (mw *MemoryWriter) Write(event string, kvs ...interface{}) {
+	var ent entry
+	ent.time = mw.clock.Now()
+	ent.event = event
+	extracted, err := extractKeyValues(kvs...)
+	if err != nil {
+		panic(err)
+	}
```

Comment:
The zapjournal writes a error message to a configurable location, currently this is set to stderr: https://github.com/filecoin-project/go-filecoin/pull/3485/files#diff-c65947145146f8d0a4bae969afded6abR18

I _really_ don't want the `Write` interface to return an error, we could do something similar to zap here or just ignore the errors, thoughts?